_posts/2024-08-19-quarkus-metaprogramming.adoc

```diff
@@ -0,0 +1,372 @@
+
+.Blocks diagram describing the implementation of the generated reflection-free Jackson serializers
+image::DeploymentVsRuntime.jpg[]
+
+== Making it optional
+
+Since it was not possible to reuse any of the heuristics used by Jackson to decide if and how a specific field should be serialized, and as discussed there could still exist some edge cases when the generated serializers produce a result that differs from what Jackson would do using reflection, we decided for now to disable this feature by default and give the possibility to users to opt-in. Considering that all the code implementing this new feature is self-contained in a single BuildStep, that is easily achievable using a https://quarkus.io/guides/writing-extensions#conditional-step-inclusion[conditional step inclusion].
+
+For this purpose it is sufficient to https://quarkus.io/guides/config-mappings[map a configuration to an interface] using the @ConfigMapping annotation
+
+[source, java]
+----
+@ConfigMapping(prefix = "quarkus.rest.jackson.optimization")
+@ConfigRoot(phase = ConfigPhase.BUILD_TIME)
+public interface JacksonOptimizationConfig {
+    @WithDefault("false")
+    boolean enableReflectionFreeSerializers();
+}
+----
+and also having an implementation of a BooleanSupplier reading that configuration
+
+[source, java]
+----
+class IsReflectionFreeSerializersEnabled implements BooleanSupplier {
+JacksonOptimizationConfig config;
+
+    	@Override public boolean getAsBoolean() {
+        	return config.enableReflectionFreeSerializers();
+    	}
+}
+----
+
+so that the BuildStep will be enabled only if this supplier returns true.
+
+[source, java]
+----
+@BuildStep(onlyIf = IsReflectionFreeSerializersEnabled.class)
+----
+
+In this way this optimization is turned off by default (note the `@WithDefault("false")` annotation on the boolean method) and can be enabled by simply adding the flag
+
+[source, properties]
+----
+quarkus.rest.jackson.optimization.enable-reflection-free-serializers=true
+----
+
+to the application.properties configuration file. With this last change, the final code of the complete BuildStep is available https://github.com/quarkusio/quarkus/blob/39ec43201a690e7895b643e6c328d9042e73ccf0/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/ResteasyReactiveJacksonProcessor.java#L374[here].
+
+== Putting it at work and measuring the results
+
+Now that all the development has been done, let’s try to put this at work with the same benchmark from where we started and check if this change comes with a real performance gain as we hoped. First of all it is necessary to enable this feature by setting the above mentioned flag in the Quarkus application.properties file. Additionally it will be useful to also set a second flag enabling Quarkus to dump the decompiled bytecode generated at deployment time
```

Comment:
```suggestion
Now that all the development has been done, let’s try to put this at work with the same benchmark from where we started and check if this change comes with a real performance gain as we hoped. First of all it is necessary to enable this feature by setting the above mentioned flag in the Quarkus `application.properties` file. Additionally it will be useful to also set a second flag enabling Quarkus to dump the decompiled bytecode generated at deployment time
```